tokio/src/sync/broadcast.rs

```diff
@@ -0,0 +1,980 @@
+    /// async fn main() {
+    ///     let (tx, _rx) = broadcast::channel(16);
+    ///
+    ///     // Will not be seen
+    ///     tx.send(10).unwrap();
+    ///
+    ///     let mut rx = tx.subscribe();
+    ///
+    ///     tx.send(20).unwrap();
+    ///
+    ///     let value = rx.recv().await.unwrap();
+    ///     assert_eq!(20, value);
+    /// }
+    /// ```
+    pub fn subscribe(&self) -> Receiver<T> {
+        let shared = self.shared.clone();
+
+        let mut tail = shared.tail.lock().unwrap();
+
+        if tail.rx_cnt == MAX_RECEIVERS {
+            panic!("max receivers");
+        }
+
+        tail.rx_cnt = tail.rx_cnt.checked_add(1).expect("overflow");
+        let next = tail.pos;
+
+        drop(tail);
+
+        Receiver {
+            shared,
+            next,
+            wait: Arc::new(WaitNode {
+                queued: AtomicBool::new(false),
+                waker: AtomicWaker::new(),
+                next: CausalCell::new(ptr::null()),
+            }),
+        }
+    }
+
+    /// Returns the number of active receivers
+    ///
+    /// An active receiver is a [`Receiver`] handle returned from [`channel`] or
+    /// [`subscribe`]. These are the handles that will receive values sent on
+    /// this [`Sender`].
+    ///
+    /// # Note
+    ///
+    /// It is not guaranteed that a sent message will reach this number of
+    /// receivers. Active receivers may never call [`recv`] again before
+    /// dropping.
+    ///
+    /// [`recv`]: crate::sync::broadcast::Receiver::recv
+    /// [`Receiver`]: crate::sync::broadcast::Receiver
+    /// [`Sender`]: crate::sync::broadcast::Sender
+    /// [`subscribe`]: crate::sync::broadcast::Sender::subscribe
+    /// [`channel`]: crate::sync::broadcast::channel
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, _rx1) = broadcast::channel(16);
+    ///
+    ///     assert_eq!(1, tx.receiver_count());
+    ///
+    ///     let mut _rx2 = tx.subscribe();
+    ///
+    ///     assert_eq!(2, tx.receiver_count());
+    ///
+    ///     tx.send(10).unwrap();
+    /// }
+    /// ```
+    pub fn receiver_count(&self) -> usize {
+        let tail = self.shared.tail.lock().unwrap();
+        tail.rx_cnt
+    }
+
+    fn send2(&self, value: Option<T>) -> Result<(), SendError<Option<T>>> {
+        let mut tail = self.shared.tail.lock().unwrap();
+
+        if tail.rx_cnt == 0 {
+            return Err(SendError(value));
+        }
+
+        // Position to write into
+        let pos = tail.pos;
+        let rem = tail.rx_cnt;
+        let idx = (pos & self.shared.mask as u64) as usize;
+
+        // Update the tail position
+        tail.pos = tail.pos.wrapping_add(1);
+
+        // Get the slot
+        let slot = &self.shared.buffer[idx];
+
+        // Acquire the write lock
+        let mut prev = slot.lock.fetch_add(1, SeqCst);
+
+        while prev & !1 != 0 {
```

Comment:
nit/tioli: This condition feels kinda confusing to me, maybe just because it has two bangs in it...